java/patterns/src/main/java/dev/restate/patterns/Compensations.java

```diff
@@ -0,0 +1,94 @@
+package dev.restate.patterns;
+
+import static dev.restate.patterns.compensations.generated.Proto.*;
+
+import dev.restate.patterns.compensations.generated.*;
+import dev.restate.sdk.RestateContext;
+import dev.restate.sdk.common.TerminalException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * -------------------------------------- Compensations --------------------------------------
+ * Restate guarantees that your invocations are run to completion no matter what happens. However,
+ * in case you want your services to be able to respond with exceptions as part of the normal
+ * control flow (i.e. denoting a negative outcome) you can let the callee throw a
+ * `TerminalException`. If the caller has completed previous steps before receiving the
+ * `TerminalException`, it will have to undo those steps in order to leave the system in a
+ * consistent state. The way to undo steps is by registering compensations and running them for all
+ * completed steps. With Restate, this can be naturally expressed as part of the service code.
+ * Moreover, Restate runs the compensations with the same guarantees as the service code, i.e. it
+ * executes them durably until they complete, which guarantees that the system will be left in a
+ * consistent state.
+ *
+ * <p>The very same mechanism of undoing completed steps via compensations is also needed if you
+ * want to gracefully cancel an invocation, i.e. because it is no longer needed or stuck, and leave
+ * the system in a consistent state. The way a graceful cancellation manifests is by throwing a
+ * `TerminalError` at the next Restate API call. In fact, Restate starts cancelling your invocation
+ * call tree starting with the leave invocations first and then propagating the cancellation error
+ * up. Given that your service code is properly instrumented with compensations, you know that an
+ * invocation which completed with a `TerminalError` has not made any changes to the system and the
+ * caller only needs to undo the steps that have been completed successfully before.
+ */
+public class Compensations {
+
+  /**
+   * Trip reservation workflow which has been instrumented with compensations. The workflow tries to
+   * reserve the flight and the car rental before it processes the payment. If at any point one of
+   * the calls fails or gets cancelled, then the trip reservation workflow will undo all
+   * successfully completed steps by running the compensations.
+   *
+   * <p>Note: that the compensation logic is purely implemented in the user code and runs durably
+   * until it completes. Moreover, an invocation failure and an invocation cancellation are handled
+   * in the exact same way by the caller.
+   */
+  public static class TravelService extends TravelRestate.TravelRestateImplBase {
```

Comment:
Perhaps how about adding a test as well? Would it be useful to showcase this scenario even more?

You could add a test that mounts this TravelService, and then mounts a mock FlightService and CarRentalService. The first succeeds, the second fails.

Reply:
Will add a small test.